### PR TITLE
Adds the pipelines:diff command

### PR DESCRIPTION
**Pull request by krummi**
This PR implements the "heroku pipelines:diff" command.

Notes on implementation:
* Got info on the undocumented kolkrabbi endpoints from a Herokai (thanks @gudmundur).
* Tried to stick to conventions in this project (i.e. regarding what to use from ES6 etc).
* Added two dependencies; lodash and request. They're probably easy to get rid of, if you guys want that.
* There are currently no tests for this, we should probably add some.

First time doing ES6, so review those parts carefully. ;)

### CODE REVIEW

File: commands/pipelines/diff.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    url: `https://api.github.com/repos/${repo}/compare/${downstreamApp.hash}...${sourceApp.hash}`,
+    headers: {
+      authorization: 'token ' + githubToken,
+      'user-agent': herokuUserAgent,
+    },
+    json: true
+  }).get(1);
+
+  console.log(`\n${sourceApp.name} is ahead of ${downstreamApp.name} by ${githubDiff.ahead_by} commit${githubDiff.ahead_by === 1 ? '' : 's'}:`);
+  for (let i = githubDiff.commits.length - 1; i >= 0; i--) {
```
Comment (appleton) [inline commands/pipelines/diff.js]: By chance I tested this on a pipeline where the commit sha of the production app is no longer available on GitHub - I think I must've rebased and force pushed or something. Anyway, that case causes an exception on this line:

```sh
TypeError: Cannot read property 'length' of undefined
    at diff (/Users/aappleton/projects/heroku-pipelines/commands/pipelines/diff.js:67:34)
    at next (native)
    at onFulfilled (/Users/aappleton/projects/heroku-pipelines/node_modules/heroku-cli-util/node_modules/co/index.js:65:19)
    at tryCatcher (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/promise.js:507:31)
    at Promise._settlePromiseAt (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/promise.js:581:18)
    at Promise._settlePromises (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/promise.js:697:14)
    at Async._drainQueue (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/async.js:123:16)
    at Async._drainQueues (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/async.js:15:14)
    at processImmediate [as _immediateCallback] (timers.js:368:17)
```

**Comment by appleton** [comment]
Thanks for the contribution! I'm making a list of cases which I'll run through to test:

- [x] App and all downstream apps are connected to GitHub
- [ ] App is connected to GitHub, some downstream apps are not
- [ ] App is connected to GitHub, all downstream apps are not
- [ ] App is not connected to GitHub

File: commands/pipelines/diff.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    const downstreamHashes = yield cli.action(`Fetching latest app releases for downstream apps`,
+        bluebird.all(downstreamApps.map(function (app) {
+          return wrappedGetLatestCommitHash(app.name, app.id);
+        })));
+
+    // Try to refrain from doing any GitHub/kolkrabbi API requests if none of the downstream
+    // hashes differ from the target hash.
+    const uniqueDownstreamHashes = _.uniq(_.pluck(downstreamHashes, 'hash'));
+    if (uniqueDownstreamHashes.length === 1 && uniqueDownstreamHashes[0] === targetHash.hash) {
+      console.log(`\nEverything is up to date.`);
```
Comment (appleton) [inline commands/pipelines/diff.js]: I think it would be good to print out which apps were compared here. This will give the user a little more feedback in the case that some of the apps they tried to compare are not connected to GitHub. How about:

```
my-app-staging is up to date with my-app-production
my-app-staging was not compared to my-app-production as my-app-production is not connected to GitHub
```

If this is tricky to implement for some reason then a simpler message is also fine but I think we should make the GitHub connection limitation clear.

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+node_modules/
```
Comment (appleton) [inline .gitignore]: :) thanks

File: commands/pipelines/diff.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    method: 'GET',
+    path: `/apps/${appId}/slugs/${release[0].slug.id}`,
+    headers: { 'Accept': V3_HEADER }
+  });
+  return { name: appName, hash: slug.commit };
+}
+
+function* diff(sourceApp, downstreamApp, repo, githubToken, herokuUserAgent) {
+  if (sourceApp.hash === downstreamApp.hash) {
+    console.log(`\neverything is up to date between ${sourceApp.name} and ${downstreamApp.name}`);
```
Comment (appleton) [inline commands/pipelines/diff.js]: Could we use `cli.log` here please? It integrates better with the test suite.

File: commands/pipelines/diff.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+function* fetchPipelineCoupling(heroku, appName) {
+  try {
+    return yield heroku.request({
+      method: 'GET',
+      path: `/apps/${appName}/pipeline-couplings`,
+      headers: { 'Accept': PIPELINES_HEADER }
+    });
+  } catch (err) {
+    console.log(err);
```
Comment (appleton) [inline commands/pipelines/diff.js]: You could change this to a `cli.debug` statement. That way it'll be hidden for most users.

File: commands/pipelines/diff.js (hunk 1 above)
Comment (krummi) [inline commands/pipelines/diff.js]: Ah, good catch!

**Comment by krummi** [comment]
Awesome - very thorough review. Will try to address the things you mention soon.

**Comment by krummi** [comment]
I've updated the PR to reflect your comments @appleton. It now looks much more carefully at the GitHub connections of the target/downstream apps as you mentioned and provides error reporting on more edge cases than before. I've also added a pretty thorough test-suite for the command.

**Comment by appleton** [comment]
This is great! Thanks so much for this and extra special thanks for adding the tests :100:

**Comment by appleton** [comment]
Published to npm as v0.27.0. Thanks again! :beers:

**Comment by krummi** [comment]
Fantastic and glad to help! :+1:
